### PR TITLE
spectral-extraction: API

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This PR extends on #1514 with a more flexible user-API for 2D to 1D spectral extraction.

[Updated spectral extraction docs](https://jdaviz--1554.org.readthedocs.build/en/1554/specviz2d/plugins.html#spectral-extraction)

- [x] requires https://github.com/astropy/specreduce/pull/117 (approved, merged, released, and pinned here), and is expected to fail CI until then.

Proposed public methods on the spectral-extraction plugin, accessed with `specviz2d.app.get_tray_item_from_name('spectral-extraction')`:
* `import_trace(trace)` - takes a specreduce trace object and loads the _input parameters_ into the plugin.  Supports `FlatTrace` and `KosmosTrace`.
* `export_trace(add_data=False, **kwargs)` - creates and returns a trace from the input parameters in the plugin.  If `add_data=True`, this adds a new data entry (once supported by glue-astronomy).
* `import_bg(bg)` - takes a specreduce background object and loads the _input parameters_ into the plugin.  Supports one-sided and two-sided backgrounds only, recognized relative to the currently set trace.
* `export_bg(**kwargs)` - creates and returns a specreduce background object from the input parameters in the plugin.
* `export_bg_img(add_data=False, **kwargs)` - returns the background 2d spectrum, optionally adding it as a data entry in the app/viewer.
* `export_bg_sub(add_data=False, **kwargs)` - returns the background-subtracted 2d spectrum, optionally adding it as a data entry in the app/viewer.
* `import_extract(ext)` - takes a specreduce extraction object and loads the _input parameters_ into the plugin.  Supports `BoxcarExtract` (should be extended to `OptimalExtract` once supported by the plugin).
* `export_extract(**kwargs)` - creates and returns a specreduce extract object from the input parameters in the plugin.
* `export_extract_spectrum(add_data=False, **kwargs)` - returns the extracted spectrum, optionally adding it as a data entry in the app/viewer.

For all `export_` methods: `**kwargs` allows setting traitlets directly from the method, and raises an error if any unsupported kwargs are passed.

For all methods: calling the method or changing a traitlet counts as an interaction with the applicable "step" and updates the visualization in real time (see screen recording below).

Note: I think we should give some thought about the import/export prefixes (these could alternatively be load/update and get, for example).  We could also consider abstracting these away with getters/setters for `trace`, `background`, and `extract` attributes, but then would lose the ability to pass `**kwargs` and so would require setting traitlets manually or through the specreduce API as well as the ability to request `add_data=True` which is particularly useful for the background and background-subtraction methods.

https://user-images.githubusercontent.com/877591/183725054-06d13233-17ef-494d-9bdb-6f754d126903.mov

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: setup.cfg
Comment (ojustino) [inline setup.cfg:37]: What's your motivation for pinning instead of setting 1.1.0 as the minimum?

**Review by ojustino — COMMENTED** [review]
I am in favor of this approach. The import/export prefixes sound best to me since that's the typical nomenclature for other Jdaviz operations that take in or emit data. I think there's value in being able to pass keyword arguments.

I know I left a bevy of comments, but most of them are about promoting interoperability with the `specreduce` and `specutils` documentation as opposed to problems with the code itself.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:402]: I don't think so (not yet at least - see https://github.com/spacetelescope/jdaviz/pull/1580#discussion_r949447571)... but updating the API docs to be ready when we do enable that mapping might still be a good idea.  I can go through and add them all though - no need for you to enter each of them manually!

File: setup.cfg
Comment (kecnry) [reply setup.cfg:37]: to avoid the situation we ran into yesterday.  Until we enforce deprecating API breaks in specreduce development, we need to ensure that an existing build of jdaviz won't be broken by a new release of specreduce.  We could probably allow for bugfixes though, with `specreduce>=1.1.0,<1.2.0`?

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:612]: I'll need to think about this one.  The combination of this PR with #1572 will mean returning either BoxcarExtract or OptimalExtract.  So maybe we should:
* implement a base Extract class in specreduce?
* do the same as for trace and point to specreduce.extract module instead?

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:452]: good catch!  I think this is just something I missed when splitting the support for importing/exporting traces from #1514 to #1556 (I removed the button that calls this method, but forgot to remove the method itself).  If #1556 gets in before this, then the `NotImplementedError` will be removed.

File: setup.cfg
Comment (ojustino) [reply setup.cfg:37]: I like that suggestion. I think `~1.1.0` means the same thing if we'd like to practice our semver-style shortcuts.

**Comment by camipacifici** [comment]
Naively, I would have expected to be able to reload any 2D spectrum I get out using `viz.load_data(spectrum2D, data_label='bla')`. Can this be an option at all? I agree that adding more methods would generate confusion.

About the bug...I will try to record the steps if/when I see it again.

**Comment by kecnry** [comment]
I do like that idea, but `specviz2d.load_data` currently assumes the first object is 2d spectrum that should be collapsed to a 1d spectrum if a 1d spectrum is not also provided, which probably would conflict with what you want to do here.  Let me think if there's a way to do that logic without a complete rewrite of the parser logic.

**Comment by camipacifici** [comment]
I was thinking of being able to feed in the background subtracted image I get out from `Background` which in principle is a 2D spectrum only not in the expected format, I guess.

**Comment by kecnry** [comment]
@ojustino - I've adopted all of your suggestions that I could, thanks!  The ones that require intersphinx mapping with specreduce will not currently work and would fail building the RTD docs, so I propose we revist those as a follow-up effort once the specreduce API docs are improved.

**Comment by kecnry** [comment]
@camipacifici - unfortunately supporting importing the spectrum1d objects for the background or background-subtracted images would probably take a refactor and change to the API on `specviz2d.load_data`.  I created a follow-up ticket for us to consider how to best do that [:cat:](https://jira.stsci.edu/browse/JDAT-2690).

File: setup.cfg
Comment (kecnry) [reply setup.cfg:38]: I actually think that `~=1.1` would include `1.2` since it doesn't consider minor version bumps as breaking (https://peps.python.org/pep-0440/#compatible-release).  For now I'll just do it the verbose way to be safe, and hopefully we can loosen up the pin down the road.

File: docs/specviz2d/plugins.rst
Comment (ojustino) [reply docs/specviz2d/plugins.rst:97]: I think this one should still work since it's `specutils` and not `specreduce`. It's used on other pages, too.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:613]: I think the second option is better for now, alongside changing mentions of "[specreduce] Extract objects" in `docs/specviz2d/plugins.rst` to something more general like "[specreduce] extraction objects" until we implement something like your first bullet in `specreduce`.

File: docs/specviz2d/plugins.rst
Comment (kecnry) [reply docs/specviz2d/plugins.rst]: woops, good catch.  Fixed!

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:453]: Since it seems like this PR is higher priority than #1556 at the moment, does it make sense to remove the method here?

File: docs/specviz2d/plugins.rst
Comment (ojustino) [inline docs/specviz2d/plugins.rst:72]:
```suggestion
To import the parameters from a specreduce Trace object, whether it's new or was exported and modified in the notebook, call:meth:`~jdaviz.configs.specviz2d.plugins.spectral_extraction.spectral_extraction.SpectralExtraction.import_trace`::
```
Consistency with the analogous line in the other sections.

**Review by ojustino — COMMENTED** [review]
Understood on the inability to link to the `specreduce` documentation. I guess it's another item on the list for us over there.

Here, I just have one more edit for the documentation and [a question](https://github.com/spacetelescope/jdaviz/pull/1554#discussion_r952762003) about the sequencing of the spectral-extraction series of pull requests. I am ready to approve once both are addressed.

File: docs/specviz2d/plugins.rst
Comment (rosteen) [inline docs/specviz2d/plugins.rst:51]: What do you think of adding a `viz.spectral_extraction` or `viz.spectral_extraction_plugin` attribute to the Specviz2D helper that is set to this, so that the user doesn't have to go through the app-level incantation?

**Review by rosteen — COMMENTED** [review]
One minor comment after reading through the code, still need to test in the notebook before I approve.

File: docs/specviz2d/plugins.rst
Comment (kecnry) [reply docs/specviz2d/plugins.rst:51]: How would you feel about re-considering #1401?  It's _slightly_ more verbose than what you suggest, but I think more friendly than the current app-level method.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py]: Ok, I removed `vue_create_trace` as well as the `add_data` argument to `export_trace` in this PR, and marked #1556 as blocked by this PR (will need to be rebased on top of this once merged, and then that commit will effectively need to be reverted in #1556 to re-introduce those APIs).

File: docs/specviz2d/plugins.rst
Comment (rosteen) [reply docs/specviz2d/plugins.rst:51]: Hmm....I'm fine with thinking about this harder in the context of 1401 and not holding up this PR.

Comment (kecnry) [reply docs/specviz2d/plugins.rst:51]: I agree that documenting any methods on `viz.app` is a bit of a slippery slope, but it is one we've already started sliding down...

**Comment by rosteen** [comment]
The things I tried seemed to work as expected, but there was one thing I had a question about. After I extracted the background, I changed the width and then imported that updated background. After I did this, the separation in the plugin changed from 10 to -10 (although the display in the viewer was identical). Is there any difference between a positive and negative separation here? I would expect it not to change when I round trip a background and don't touch the separation.

**Comment by kecnry** [comment]
@rosteen - was it a one-sided or two-sided background?  If it's two sided, there is no difference, but I can easily force the positive separation by wrapping with `abs()`.  Under-the-hood, the specreduce background object actually doesn't store the separation, so they are re-estimated from the traces themselves.

**Comment by rosteen** [comment]
> @rosteen - was it a one-sided or two-sided background?

It was two-sided. I could see users getting confused if the sign flips arbitrarily, so 👍 to wrapping it in `abs`.

**Review by rosteen — APPROVED** [review]

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (ojustino) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py:601]: This line slipped through the cracks, but I modified the suggestion to match what we agreed to elsewhere. Since it's a small change, I will go ahead and give the approval so you can merge after you've made it.

**Review by ojustino — APPROVED** [review]
The tests are passing and we've covered almost all of my currently addressable comments, so I am good with a merge once [this change](https://github.com/spacetelescope/jdaviz/pull/1554/files#r952929752) is in.

File: jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py
Comment (kecnry) [reply jdaviz/configs/specviz2d/plugins/spectral_extraction/spectral_extraction.py]: thanks, I guess I just fixed the ones in plugins.rst and forgot the one that you actually had the suggestion.  Fixed now.

**Comment by kecnry** [comment]
Thanks for the reviews/suggestions!

@PatrickOgle is in the process of looking at the API, so even though this technically has 2 dev approvals, let's hold off on merging.

**Review by PatrickOgle — APPROVED** [review]
Tried all API methods on real data and they work.

**Comment by kecnry** [comment]
Thanks everyone for the reviews and comments!

**Comment by eteq** [comment]
Sorry I'm late to the party - was on vacation for most of the review period.  I mostly love this, but I had a few things  I've put in two follow-on issue: #1600, and #1601 - see those for more details.  (Note also the two issues are in some ways in conflict with naming conventions, although they could be two independent steps as well)

**Pull request merged**
